server/study_server.js

```diff
@@ -153,56 +153,77 @@ module.exports = {
     },
     submitResult: function (req, res, next) {
         var clean_id = sanitize(req.body.id);
+        var max_retries = 3;
 
-         Study.findOne({_id: clean_id}, function (err, study) {
-            if (err) {
-                res.status(504);
-                logger.error("study_server.js: Error in submit result:", err);
-                res.end(err);
-            } else {
-                //if the study is being previewed, don't record the response
-                if (req.body.resid === "preview"){
-                    res.redirect('/studies');
+        // function to save complete response to study and remove from incomplete responses
+        function saveStudy(study, responseID, retries){
+            var respIdx = study.incompleteResponses.indexOf(responseID);
+            study.completeResponses.push(study.incompleteResponses[respIdx]);
+            study.incompleteResponses.splice(respIdx,1);
```

Comment:
Would moving the response to the `completeResponses` list be best done on a successful save, instead of before trying to save?

Reply:
I think that would also work. So `saveStudy` function will look something like this:
```
function saveStudy(study, responseID, retries) {
    study.save(function(err) {
        if (err) {
            if (retries < max_retries) { .... } 
           else { ... }
        } else {
            var respIdx = study.incompleteResponses.indexOf(responseID);
            study.completeResponses.push(study.incompleteResponses[respIdx]);
            study.incompleteResponses.splice(respIdx, 1);
            logger.info("study_server.js: Response saved successfully to study.");
            res.redirect('/msg/thanks');
            res.end();
        }
    });
```

Reply:
Could you elaborate more? I thought I initially understood what you meant but I'm not too sure. Is this saving the response to the response table or saving the `responseId `to the `completeresponses` in the study table

Reply:
I'm talking about the whole update of the state of completeResponsess and incompleteResponses. As far as I can see it is being done before a successful save happens. So if the retries all fail, the state has still been updated, but no save has happened.

Reply:
Ah. Yes, the in-memory state would be different to the database if it fails all the retries.  So, if we were to change it where the arrays are only updated when a successful save occurs, I guess it would look something like this : 
```
function saveStudy(study, responseID, retries) {
   study.save(function(err) {
        if(err) { // retry logic  }
        else {
          // update array logic
         study.save()
         }
      }
}
```

Reply:
Though, what if it fails on that save.

Reply:
Is there no point at which you now the save succeeded?

Reply:
Also, I would probably move the updating of completeResponses and incompleteResponses into a separate function with a meaningful name so as to not obscure the logic of the retry function.

Reply:
Isn't the state itself that we are saving?

Reply:
Yes. you are correct and you should ignore what I've said.